comby: init at 0.12.1

Br1ght0ne opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change
https://github.com/comby-tools/comby/releases/tag/0.12.1

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/nixos/nixpkgs/76828)
<!-- Reviewable:end -->

Br1ght0ne committed:
comby: init at 0.12.1

Br1ght0ne committed:
ocamlPackages.hack_parallel: init at 0.1.1

yegortimoshenko commented on pkgs/development/ocaml-modules/hack_parallel/default.nix:
```diff
@@ -0,0 +1,32 @@
+
+  buildInputs = [ core ppx_deriving ppxlib sexplib ];
+
+  patches = [ ./dune-static-libs.patch ];
```
Nitpick: technically it's a diff, not a patch :) Compare:

- https://github.com/NixOS/nixpkgs/commit/228627d50e1be526aa87a97c282ecf9dcbacc6eb.diff
- https://github.com/NixOS/nixpkgs/commit/228627d50e1be526aa87a97c282ecf9dcbacc6eb.patch.

lukateras reviewed (commented):
Doesn't build:

```
(cd _build/default && .ppx/3a48ee47a0ff6f2bd43558b49bc4d853/ppx.exe --conditional --cookie 'library-name="server_types"' -o lib/server/server_types.pp.ml --impl lib/server/server_types.ml --dump-ast)
.ppx/3a48ee47a0ff6f2bd43558b49bc4d853/ppx.exe: unknown option '--conditional'.
builder for '/nix/store/y0wq6bdvfgywg2hcna23ix6x5g7q620f-ocaml4.07.1-comby-0.12.1.drv' failed with 
exit code 1
error: build of '/nix/store/y0wq6bdvfgywg2hcna23ix6x5g7q620f-ocaml4.07.1-comby-0.12.1.drv' failed
```

yegortimoshenko commented:
@GrahamcOfBorg build comby

JohnAZoidberg commented:
Still interested!

kvtb commented:
There is upstream interest to get it packed in `nixpkgs` (likely the author is Nix/Darwin user)
https://github.com/comby-tools/comby/issues/142

JohnAZoidberg committed:
rename patch to diff

JohnAZoidberg committed:
comby, hack_parallel: Use Dune2

File "/nix/store/5j7s3mgkvx6l0cgk7wvrbzznfq4msbaw-ocaml4.10.0-core_kernel-0.14.0/lib/ocaml/4.10.0/site-lib/core_kernel/dune-package", line 1, characters 11-14:
1 | (lang dune 2.8)
               ^^^
Error: Version 2.8 of dune is not supported.
Supported versions:
- 0.0
- 1.0 to 1.11
Done: 0/0 (jobs: 0)

JohnAZoidberg committed:
mparser-comby: init at 8de1895f6f4e48e19280ec37f3e

JohnAZoidberg commented:
I rebased it onto master and fixed a few things:

- Rename the patch file to diff
- Enable `useDune2 = true` because that's apparently required now.

And because comby depends on a few custom dependencies we need to either push upstream to upstream their forked dependency changes or patch/package those (see https://github.com/comby-tools/comby/blob/master/comby.opam):

- [x] mparser-comby: I temporarily added a new file for this
- [ ] patdiff: Requires a different branch that doesn't cleanly merge with upstream, so I can't create a patch file (https://github.com/rvantonder/patdiff.git#0.13.0-patch-compatible-diffs)
- [ ] hack_parallel: Not sure if this is the correct version
- [ ] bisect_ppx: Not sure if this causes issues
- [ ] camlzip: Not sure if this causes issues

kvtb commented:
`1.0.0` is released, `hack_parallel` can be trivially updated, but `comby` is tricky

dywedir commented:
Thanks!
Closing the PR, please see https://github.com/NixOS/nixpkgs/pull/123986, https://github.com/NixOS/nixpkgs/pull/139408
